### DISCUSSION TITLE
Configuration appearing to be ignored in docker

### DISCUSSION THREAD

**Discussion by MattInternet** [bluenviron/mediamtx#278] [Questions]
## Which version are you using?

v0.13.2

## Which operating system are you using?

OS

- [x] Linux
- [ ] Windows
- [ ] macOS

Architecture

- [x] amd64
- [ ] arm64
- [ ] arm7
- [ ] arm6

## Describe the issue

I am mounting a volume with a configuration in it that works locally. Once the volume is mounted it appears to be picked up because the container runs and stars putting out logs. It also says `reloading configuration` if I edit the config, so im quite sure it is getting the configuration. When I connect a stream to it it says connected and `is publishing to path` and everything looks fine except it isnt running my command in `runOnPublish`

Its strange because if I just download the binary and run it it with this same config it works perfectly. The current content in `runOnPublished` uses ffmpeg to output a file but it also does nothing in the container if i replace it with very simple things like `echo 'cat'`

## Describe how to replicate the issue

1. Make a directory and put the attached `config.txt`file in it (rename it to `config.yml` because GitHub wont let me attach a `.yml` file. [config.txt](https://github.com/aler9/rtsp-simple-server/files/5817516/config.txt)
2. `cd` into the directory and run `docker run --rm -it -e RTSP_PROTOCOLS=tcp -v ${PWD}:/app -w /app -p 8554:8554 aler9/rtsp-simple-server ./config.yml`

**Comment by MattInternet** [comment]
Super awesome project btw, thank you so much :D

**Comment by aler9** [comment]
Hello,

1. I've added to the README the right command to replace the configuration file when using Docker
    https://github.com/aler9/rtsp-simple-server#configuration

2. anyway, this is not your issue: the issue is that you're trying to use ffmpeg, but the Docker image doesn't contain ffmpeg, since there would be license issues and the image would become huge. You have to generate an image that contains both rtsp-simple-server and ffmpeg, in this way

```
FROM aler9/rtsp-simple-server AS rtsp
FROM alpine:3.12
RUN apk add --no-cache ffmpeg
COPY --from=rtsp /mediamtx /
COPY --from=rtsp /mediamtx.yml /
ENTRYPOINT [ "/mediamtx" ]
```

**Reply by rmwiseman** [reply]
Thanks for this.

In case I'm not the only person who didn't know exactly what to do with it, here's what I did after a bit of googling:
1. Put the text above (`FROM aler9/rtsp-simple-server AS rtsp` etc.) into a file called `Dockerfile`.
2. In the same directory as where `Dockerfile` is, build it, giving it whatever tag you choose (I used `rss-ffmpeg:latest` because I'm a lazy typist): `docker build -t rss-ffmpeg:latest .` (the dot at the end is important!)
3. Then you'll see it when you run `docker image ls` and can run it using `docker run rss-ffmpeg:latest` (along with all the appropriate switches - [see here](https://github.com/aler9/rtsp-simple-server#docker), but use your tag instead of `aler9/rtsp-simple-server`).

**Reply by ghost** [reply]
Thanks for the clear explanation! 

However I have one doubt: **As you are now in a Linux distribution (Alpine), what command do you pass with _ffmpeg_ to get video from an input device?**

According the documentation I should include in the configuration file the following command: _ffmpeg -f v4l2 -i /dev/video0_ ... 
But I cannot find the _/dev/video0_ folder in the container and when I run the command _ffmpeg -devices_ I obtain: 

>  DE alsa            ALSA audio output
>  DE fbdev           Linux framebuffer
>  D  lavfi           Libavfilter virtual input device
>  DE oss             OSS (Open Sound System) playback
>   E sdl,sdl2        SDL2 output device
>  DE video4linux2,v4l2 Video4Linux2 output device
>  D  x11grab         X11 screen capture, using XCB

**Reply by Tuxie** [reply]
You'll need to add `--device=/dev/video0` to the docker command line, or a `devices:` section if you use docker-compose.

**Reply by Tuxie** [reply]
I assumed that you had it working on the host since you wrote that you didn't have it _in the container_.
Before it can work in the container you must have v4l2 working on the host. I assume it's a local (USB or builtin) webcam that you want to stream? If it isn't detected automatically you may have to load some driver and `modprobe v4l2-common` or similar. Unfortunately I can't help you with this because it was something like 2 decades ago I dealt with v4l2 last.

**Reply by ghost** [reply]
When running _docker run --rm -it --device=/dev/video0_... 
I obtain: _Error response from daemon: error gathering device information while adding custom device "/dev/video0": not a device node._

My host is Windows

**Reply by Tuxie** [reply]
I may be wrong but I don't think it's possible to access raw devices from within WSL (1 or 2), unfortunately. You'll have to run ffmpeg outside of docker and have it publish the stream to the rtsp-simple-server running in docker. See https://trac.ffmpeg.org/wiki/DirectShow and as output use something like `-f rtsp rtsp://<rtsp-streaming-server's ip>:<rtsp port>/my-webcam` and then you can access it like any other rtsp stream from there.

**Reply by ghost** [reply]
I can't find a way to make it work but thank you for your help.
I followed your advice and I have launched a rtsp producer outside the docker container (rtsp://localhost:8554/cam), however I don't find a way to consume the stream inside the docker container...

**Reply by Tuxie** [reply]
Once the outside ffmpeg is producing, you should be able to consume it the same way inside the container (try rtsp://127.0.0.1:8554/cam)

**Reply by ghost** [reply]
So inside the container if I run a python script with OpenCV, can I get the video?

```
import cv2
import os

os.environ["OPENCV_FFMPEG_CAPTURE_OPTIONS"] = "rtsp_transport;udp"
vcap = cv2.VideoCapture("rtsp://localhost:8554/cam", cv2.CAP_FFMPEG)
while(1):
    ret, frame = vcap.read()
    if ret == False:
        print("Frame is empty")
        break;
    else:
        cv2.imshow('VIDEO', frame)
        cv2.waitKey(1)
```

Will there be no flow problems? This code works on the host...

**Reply by Tuxie** [reply]
Yes, you should be able to run this inside the same container that is running rtsp-simple-server. If you run it in a separate container you need to replace "localhost" with the ip of the rtsp-simple-server container.

**Reply by ghost** [reply]
Hmmm I didn't thought about that... So I would have to run on the producer's terminal (Windows host): _ipconfig /all_ and identify the one that says _IPv4 address_?
Thank you very much for the help! It has been of great value

**Reply by aler9** [reply]
>  I don't think it's possible to access raw devices from within WSL (1 or 2)

Exactly. Docker on Windows uses a Linux virtualization layer (WSL) that doesn't expose most of the devices (GPU, devices, etc). Use the native version or use VirtualBox/VMware, that have better hardware support.

**Reply by bartowski1182** [reply]
I must be crazy, I can't get this Dockerfile to build. Every time I get an error at the COPY that /rtsp-simple-server doesn't exist:

 > [stage-1 3/3] COPY --from=rtsp /rtsp-simple-server /:
------
Dockerfile:4
--------------------
   2 |     FROM alpine:latest
   3 |     RUN apk add --no-cache ffmpeg
   4 | >>> COPY --from=rtsp /rtsp-simple-server /
   5 |     ENTRYPOINT [ "/rtsp-simple-server" ]
   6 |
--------------------
ERROR: failed to solve: failed to compute cache key: failed to calculate checksum of ref moby::xfr910rkjvqzsu5993k99ru85: "/rtsp-simple-server": not found

**Reply by aler9** [reply]
@noneabove1182 the executable was renamed into /mediamtx since v0.22.0, so you have to change /rtsp-simple-server into /mediamtx

**Comment by AndBondStyle** [comment]
@aler9 maybe you can CI-build an alternative docker tag with ffmpeg included? That would be very convenient

**Comment by Tuxie** [comment]
FYI, in addition to `ffmpeg` I also added `libva` and `mesa-va-gallium` to the Dockerfile to get amd/radeonsi vaapi hardware acceleration working. Used the docker instructions from Frigate to add the `/dev/dri/renderD128` device and set `LIBVA_DRIVER_NAME=radeonsi`

**Reply by AdiAbuAli** [reply]
can you please share the syntax in docker file to enable HW accelerator?

**Reply by Tuxie** [reply]
This is my full Dockerfile for AMD GPU hardware acceleration. For Intel or Nvidia you will need some other packages.
```
FROM aler9/rtsp-simple-server AS rtsp
FROM alpine:latest
RUN apk add --no-cache libva mesa-va-gallium ffmpeg
COPY --from=rtsp /rtsp-simple-server /
ENTRYPOINT [ "/rtsp-simple-server" ]
```
Then in my docker-compose.yml I have
```
    privileged: true
    devices:
      - /dev/dri/renderD128
    environment:
      LIBVA_DRIVER_NAME: "radeonsi"
```
and I believe the docker cli equivalents are `--privileged --device /dev/dri/renderD128 --env LIBVA_DRIVER_NAME=radeonsi`

**Comment by AdiAbuAli** [comment]
Thanks ....What was the FFMPEG command you used?

**Comment by Tuxie** [comment]
For example this one to transcode a large high bitrate video to a smaller low bitrate mobile version:
```
  baksidan-mobile:
    runOnDemand:
      ffmpeg -hide_banner -loglevel info
      -hwaccel vaapi -hwaccel_output_format vaapi -hwaccel_device /dev/dri/renderD128
      -avoid_negative_ts make_zero -fflags +genpts+discardcorrupt -flags low_delay
      -stimeout 5000000 -rtsp_transport tcp
      -i "rtsp://127.0.0.1/baksidan"
      -vf "scale_vaapi=w=1280:h=720" -c:a copy -c:v h264_vaapi -b:v 768K
      -f rtsp rtsp://127.0.0.1/$RTSP_PATH
    runOnDemandRestart: yes
```

**Reply by AdiAbuAli** [reply]
Thank you dear..I tried allpossible configurations but all failed with device creation failure -12
I opened question thread in Synology

**Reply by gLuColte** [reply]
Sorry where does this go to?

I am trying to build a single docker that streams local file (video within rtsp-simple-server docker) to a port.

Can this work?

What I want to achieve is a single docker that incooperates rtsp-simple-server and ffmpeg, incooperating with a shell script that fetches video from online storage bucket.

**Comment by cedricve** [comment]
When using this I'm getting a:

    ffmpeg: error while loading shared libraries: libblas.so.3: cannot open shared object file: No such file or directory
    2022/11/26 15:14:15 INF [path usbcam] runOnInit command exited with code 127

[a comment by an automated account is not shown]
